src/main/java/com/zendesk/maxwell/bootstrap/SynchronousBootstrapper.java

```diff
@@ -0,0 +1,236 @@
+			int insertedRows = 0;
+			while ( resultSet.next() ) {
+				RowMap row = new RowMap(
+						"insert",
+						databaseName,
+						tableName,
+						System.currentTimeMillis(),
+						table.getPKList(),
+						position);
+				setRowValues(row, resultSet, table);
+
+				if ( LOGGER.isDebugEnabled() )
+					LOGGER.debug("bootstrapping row : " + row.toJSON());
+
+				producer.push(row);
+				++insertedRows;
+				updateInsertedRowsColumn(insertedRows, startBootstrapRow, position, connection);
+			}
+			setBootstrapRowToCompleted(insertedRows, startBootstrapRow, connection);
+		}
+		catch ( NoSuchElementException e ) {
+			LOGGER.info(String.format("bootstrapping cancelled for %s.%s", databaseName, tableName));
+		}
+	}
+
+	private void updateInsertedRowsColumn(int insertedRows, RowMap startBootstrapRow, BinlogPosition position, Connection connection) throws SQLException, NoSuchElementException {
+		long now = System.currentTimeMillis();
+		if ( now - lastInsertedRowsUpdateTimeMillis > INSERTED_ROWS_UPDATE_PERIOD_MILLIS ) {
+			long rowId = ( long ) startBootstrapRow.getData("id");
+			String sql = "update maxwell.bootstrap set inserted_rows = ?, binlog_file = ?, binlog_position = ? where id = ?";
+			PreparedStatement preparedStatement = connection.prepareStatement(sql);
+			preparedStatement.setInt(1, insertedRows);
+			preparedStatement.setString(2, position.getFile());
+			preparedStatement.setLong(3, position.getOffset());
+			preparedStatement.setLong(4, rowId);
+			if ( preparedStatement.executeUpdate() == 0 ) {
+				throw new NoSuchElementException();
+			}
+			lastInsertedRowsUpdateTimeMillis = now;
+		}
+	}
+
+	protected Connection getConnection( ) throws SQLException {
+		return context.getConnectionPool().getConnection();
+	}
+
+	private RowMap bootstrapStartRowMap(Table table, BinlogPosition position) {
+		return bootstrapEventRowMap("bootstrap-start", table, position);
+	}
+
+	private RowMap bootstrapCompleteRowMap(Table table, BinlogPosition position) {
+		return bootstrapEventRowMap("bootstrap-complete", table, position);
+	}
+
+	private RowMap bootstrapEventRowMap(String type, Table table, BinlogPosition position) {
+		return new RowMap(
+				type,
+				table.getDatabase().getName(),
+				table.getName(),
+				System.currentTimeMillis(),
+				table.getPKList(),
+				position);
+	}
+
+	@Override
+	public void completeBootstrap(RowMap completeBootstrapRow, AbstractProducer producer, MaxwellReplicator replicator) throws Exception {
+		String databaseName = ( String ) completeBootstrapRow.getData("database_name");
+		String tableName = ( String ) completeBootstrapRow.getData("table_name");
+		Database database = findDatabase(replicator.getSchema(), databaseName);
+		ensureTable(tableName, database);
+		Table table = findTable(tableName, database);
+		OpenReplicator openReplicator = replicator.getOpenReplicator();
+		BinlogPosition position = new BinlogPosition(openReplicator.getBinlogPosition(), openReplicator.getBinlogFileName());
```

Comment:
oh, this isn't quite correct.  the open-replicator's binlog position may be ahead of maxwell's, as it runs in an asynchronous thread.  What you really want to do is attach a `BinlogPosition` object to the `RowMap`.